### PR TITLE
Fix data races in lazy construction and show() 

### PR DESCRIPTION
**Pull request by NHDaly**
Need to always lock accesses to base_cache :)

- Forgot to lock during lazy construction in thread_cache(tid)
- Added a lock+copy to show(), to avoid holding the lock while printing
  to IO, which may block arbitrarily long
    - Perf improvement for Base.show: print to buffer instead of copy

### CODE REVIEW

**Comment by NHDaly** [comment]
(Leaving some breadcrumbs: This PR was informally reviewed by @tkf over slack 😊 Thanks again @tkf!!)

File: src/MultiThreadedCaches.jl [hunk 1]
```diff
@@ -151,10 +185,10 @@ function Base.get!(func::Base.Callable, cache::MultiThreadedCache{K,V}, key) whe
         Base.@lock tlock begin
             if test_haskey
                 if !haskey(tcache, key)
-                    setindex!(tcache, key, v)
+                    tcache[key] = v
                 end
             else
-                setindex!(tcache, key, v)
+                tcache[key] = v
```
Comment (NHDaly) [inline src/MultiThreadedCaches.jl:191]: 😅 looks like we don't have good test coverage.. the key and value were backwards in these calls. i always screw that up with setindex 😅

Comment (NHDaly) [reply src/MultiThreadedCaches.jl:191]: Oh, haha, it's because i was using `Dict{Int,Int}` everywhere in the tests so i never noticed 😂 😭 😂

**Review by vilterp — APPROVED** [review]

**Comment by vilterp** [comment]
the two `show`s seem almost the same, but I guess just different enough to prevent DRYing 🤷‍♂️ 

never seen one with a mime type before; that's cool! Guess that's what `display` is for?

File: src/MultiThreadedCaches.jl (hunk 1 above)
Comment (NHDaly) [reply src/MultiThreadedCaches.jl:191]: K, added a test for this in 073cbc5.

**Comment by NHDaly** [comment]
yeah, exactly. The 3-arg show with the mime type is what prints at the repl if you just do:
```julia
julia> cache
MultiThreadedCache: Dict{Any, Any} with 1 entry:
  1 => "hi"
```
vs 2-arg show, which you can either call directly, or which also gets called if the value is inside another container:
```julia
julia> show(cache)
MultiThreadedCache{Any, Any}(Dict{Any, Any}(1 => "hi"))
julia> julia> [cache, cache]
2-element Vector{MultiThreadedCache{Any, Any}}:
 MultiThreadedCache{Any, Any}(Dict{Any, Any}(1 => "hi"))
 MultiThreadedCache{Any, Any}(Dict{Any, Any}(1 => "hi"))
```

**Pull request merged**